src/core/Akka.Remote/Transport/Streaming/StreamAssociationHandle.cs

```diff
@@ -0,0 +1,238 @@
+        }
+
+        private async Task ReadLoop()
+        {
+            //TODO Add 100% test coverage for this
+
+            byte[] readBuffer = _readBuffer;
+            int bufferLength = readBuffer.Length;
+
+            int readBytes = 0;
+            int readIndex = 0;
+
+            while (true)
+            {
+                int available = readBytes - readIndex;
+                if (available < sizeof(int))
+                {
+                    // To simplify the message length read logic, we enforce the length bytes be whole in the buffer
+
+                    // Copy the partial length at the start of the buffer
+                    if (available != 0)
+                        Buffer.BlockCopy(readBuffer, readIndex, readBuffer, 0, available);
+
+                    readIndex = 0;
+                    readBytes = await _stream.ReadAsync(readBuffer, available, bufferLength - available).ConfigureAwait(false);
+
+                    if (readBytes == 0)
+                    {
+                        _eventListener.Notify(new Disassociated(DisassociateInfo.Unknown));
+                        return;
+                    }
+
+                    // Adjust readBytes to include the partial length that we copied earlier
+                    readBytes += available;
+                }
+
+                int payloadLength = BitConverter.ToInt32(readBuffer, readIndex);
+                readIndex += sizeof(int);
+
+                //TODO Protect against invalid payload length by chunking the payload buffer
+                //     if payloadLength is bigger than available data in the read buffer
+                byte[] payloadBuffer = new byte[payloadLength];
+                int payloadOffset = 0;
+
+                while (payloadOffset < payloadLength)
+                {
+                    if (readIndex == readBytes)
+                    {
+                        readIndex = 0;
+                        readBytes = await _stream.ReadAsync(readBuffer, 0, bufferLength).ConfigureAwait(false);
+
+                        if (readBytes == 0)
+                        {
+                            _eventListener.Notify(new Disassociated(DisassociateInfo.Unknown));
+                            return;
+                        }
+                    }
+
+                    available = Math.Min(readBytes - readIndex, payloadLength - payloadOffset);
+
+                    Buffer.BlockCopy(readBuffer, readIndex, payloadBuffer, payloadOffset, available);
+                    readIndex += available;
+                    payloadOffset += available;
+                }
+
+                var payload = new InboundPayload(ByteString.Unsafe.FromBytes(payloadBuffer));
+                _eventListener.Notify(payload);
+            }
+        }
+
+        public override bool Write(ByteString payload)
+        {
+            _writeQueue.Enqueue(payload);
+
+            // Always return true, we don't want to use the backoff in EndpointWriter
```

Comment:
Using backoff might actually be the right thing to do in some cases, like if the writeQueue is full.